python/tests/test_types.py

```diff
@@ -204,3 +206,388 @@ def test_non_parameterized_type_equality(input_index, input_type, check_index, c
+
+
+def test_repr_double():
+    assert repr(DoubleType()) == "DoubleType()"
+
+
+def test_serialization_date():
+    assert DateType().json() == '"date"'
+
+
+def test_deserialization_date():
+    assert TestType.parse_raw('"date"') == DateType()
+
+
+def test_str_date():
+    assert str(DateType()) == "date"
+
+
+def test_repr_date():
+    assert repr(DateType()) == "DateType()"
+
+
+def test_serialization_time():
+    assert TimeType().json() == '"time"'
+
+
+def test_deserialization_time():
+    assert TestType.parse_raw('"time"') == TimeType()
+
+
+def test_str_time():
+    assert str(TimeType()) == "time"
+
+
+def test_repr_time():
+    assert repr(TimeType()) == "TimeType()"
+
+
+def test_serialization_timestamp():
+    assert TimestampType().json() == '"timestamp"'
+
+
+def test_deserialization_timestamp():
+    assert TestType.parse_raw('"timestamp"') == TimestampType()
+
+
+def test_str_timestamp():
+    assert str(TimestampType()) == "timestamp"
+
+
+def test_repr_timestamp():
+    assert repr(TimestampType()) == "TimestampType()"
+
+
+def test_serialization_timestamptz():
+    assert TimestamptzType().json() == '"timestamptz"'
+
+
+def test_deserialization_timestamptz():
+    assert TestType.parse_raw('"timestamptz"') == TimestamptzType()
+
+
+def test_str_timestamptz():
+    assert str(TimestamptzType()) == "timestamptz"
+
+
+def test_repr_timestamptz():
+    assert repr(TimestamptzType()) == "TimestamptzType()"
+
+
+def test_serialization_string():
+    assert StringType().json() == '"string"'
+
+
+def test_deserialization_string():
+    assert TestType.parse_raw('"string"') == StringType()
+
+
+def test_str_string():
+    assert str(StringType()) == "string"
+
+
+def test_repr_string():
+    assert repr(StringType()) == "StringType()"
+
+
+def test_serialization_uuid():
+    assert UUIDType().json() == '"uuid"'
+
+
+def test_deserialization_uuid():
+    assert TestType.parse_raw('"uuid"') == UUIDType()
+
+
+def test_str_uuid():
+    assert str(UUIDType()) == "uuid"
+
+
+def test_repr_uuid():
+    assert repr(UUIDType()) == "UUIDType()"
+
+
+def test_serialization_fixed():
+    assert FixedType(22).json() == '"fixed[22]"'
+
+
+def test_deserialization_fixed():
+    fixed = TestType.parse_raw('"fixed[22]"')
+    assert fixed == FixedType(22)
+
+    inner = fixed.__root__
+    assert isinstance(inner, FixedType)
+    assert inner.length == 22
+
+
+def test_str_fixed():
+    assert str(FixedType(22)) == "fixed[22]"
+
+
+def test_repr_fixed():
+    assert repr(FixedType(22)) == "FixedType(length=22)"
+
+
+def test_serialization_binary():
+    assert BinaryType().json() == '"binary"'
+
+
+def test_deserialization_binary():
+    assert TestType.parse_raw('"binary"') == BinaryType()
+
+
+def test_str_binary():
+    assert str(BinaryType()) == "binary"
+
+
+def test_repr_binary():
+    assert repr(BinaryType()) == "BinaryType()"
+
+
+def test_serialization_decimal():
+    assert DecimalType(19, 25).json() == '"decimal(19, 25)"'
```

Comment:
Isn't this type invalid because the scale is greater than the precision? (I think we check that in Java)

Reply:
Good one, I've created an issue https://github.com/apache/iceberg/issues/5038